src/internet_identity/src/state/flow_states.rs

```diff
@@ -0,0 +1,131 @@
+use crate::anchor_management::registration::Base64;
+use crate::MINUTE_NS;
+use candid::Principal;
+use ic_cdk::api::time;
+use internet_identity_interface::internet_identity::types::Timestamp;
+use std::collections::{HashMap, VecDeque};
+
+// Expiration for a registration flow, the same as the captcha expiry
+const FLOW_EXPIRATION_NS: u64 = 5 * MINUTE_NS;
+
+#[derive(Default)]
+pub struct FlowStates {
+    /// In-progress registration flows tied to the respective principal. The principal will become
+    /// a temp key upon successful completion of the registration flow.
+    /// Registration flows are removed on three conditions:
+    /// * The registration flow is completed successfully
+    /// * Once the flow expires
+    /// * The Internet Identity canister is upgraded
+    ///
+    /// The total number of flow states is limited by the registration rate limit.
+    registration_flows: HashMap<Principal, RegistrationFlowState>,
+
+    /// Deque to efficiently prune expired registration flows
+    ///
+    /// The total number of flow state expirations is limited by the registration rate limit.
+    expirations: VecDeque<FlowExpiration>,
+}
+
+/// State of a registration flow. The state expresses which action is expected next.
+#[derive(Debug, Clone)]
+pub enum RegistrationFlowState {
+    /// Client is expected to supply the `captcha_solution` using
+    /// `check_captcha`.
+    CheckCaptcha {
+        flow_created_timestamp_ns: u64,
+        captcha_solution: String,
+        captcha_png_base64: Base64,
+    },
+    /// Client is expected to finish the registration (supply an [authn_method](internet_identity_interface::internet_identity::types::AuthnMethodData))
+    /// using `identity_registration_finish`.
+    FinishRegistration { flow_created_timestamp_ns: u64 },
+}
+
+impl RegistrationFlowState {
+    pub fn create_at_timestamp_ns(&self) -> u64 {
```

Comment:
With this names it seems it's creating it.

How about `created_at_timestamp_ns`? with `created` instead of `create`.

Reply:
Ah, yes. That is a typo. Thanks!